drivers: clock_control: Add API to set the clock rate

ceolin opened the pull request:
~Add a new API to set the clock rate synchronously and asynchronously.~

v2:
 - As discussed in the last PM meeting, the async API was removed and only synchronous API is implemented now.

Commit:
drivers: clock_control: Add API to set the clock rate

Add a new API to set the clock rate synchronously and asynchronously.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

ceolin commented:
@teburd ^

teburd reviewed (approved):
I think this is fine, at some point we'll need a way to have things get notified on clock changes, but that's probably true for clock on/off as well, good stuff!

nordic-krch commented on include/drivers/clock_control.h:
```diff
@@ -231,6 +236,43 @@ static inline int clock_control_get_rate(const struct device *dev,
+ *
+ * @param dev Device structure whose driver controls the clock.
+ * @param sys Opaque data representing the clock.
+ * @param rate Subsystem clock rate.
```
is this in Hz or opaque, vendor specific value?

ceolin replied:
I believe it is easier to use if it is vendor specific.

ceolin replied:
opinions ?

nordic-krch replied:
I also think that this could be vendor specific, So maybe rephrase to `Opaque data representing clock rate`

teburd replied:
Vendor specific makes sense to me as well

ceolin replied:
cool ! Thanks for the feedback @teburd.

@nordic-krch can you take a look on it again ?

Commit:
drivers: clock_control: Opaque type for clock rate

Add an opaque type that is platform specific to be used when setting a
clock rate.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>
